File: py/classic/dynamic_programming/KnapsackOfZeroOne.py
#!/usr/bin/env python
# -*- coding:gbk -*-

from typing import List

class KnapsackOfZeroOne:
    def solution(self, weights: List[int], values: List[int], target: int) -> int:
        result = [0 for _ in range(target + 1)]
        for weight, value in zip(weights, values):
            for i in range(target, weight - 1, -1):
                pivot = result[i - weight] + value
                if result[i] < pivot:
                    result[i] = pivot
        return result[-1]

File: c/classic/dynamic_programming/KnapsackOfZeroOne.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int knapsackOfZeroOne(int *weights, int *values, int length, int target) {
    int *result = (int *) malloc(sizeof(int) * (target + 1));
    memset(result, 0, sizeof(int) * (target + 1));
    for (int i = 0; i < length; ++i) {
        int weight = weights[i];
        int value = values[i];
        for (int j = target; j >= weight; --j) {
            int pivot = result[j - weight] + value;
            if (result[j] < pivot) {
                result[j] = pivot;
            }
        }
    }
    return result[target];
}
